resources/type-system/subtyping.md

```diff
@@ -247,10 +283,71 @@ Right union introduction:
  - `S <: T0 | T1` if `S <: T0` or `S <: T1`
 
 Left intersection introduction:
- - `X & S <: T` if `X <: T` or `S <: T`
+ - `S0 & S1 <: T` if `S0 <: T` or `S1 <: T`
 
 Right intersection introduction:
- - `S <: X & T` if `S <: X` and `S <: T`
+ - `S <: T0 & T1` if `S <: T0` and `S <: T1`
+
+Axioms for `Null` and `Object`:
+
+Left non-nullable `Null`:
+ - `Null! <: T`
+
+Right nullable `Object`:
+ - `T <: Object?`
+
+General distributivity of intersection over union would also give the following:
+  - `(S0 | S1) & (S2 | S3) <: (S0 & S2) | (S0 & S1) | (S1 & S2) | (S1 & S3)`
+
+General distributivity of union over intersection would also give the following:
+  - `(S0 & S1) | (S2 & S3) <: (S0 | S2) & (S0 | S1) & (S1 | S2) & (S1 | S3)`
+
+However, we do not distribute over `FutureOr` since we cannot represent the
+result, and so we don't start from the general rules, but instead use them to
+justify specific distributivity properties of `!` and `?`.
+
+Left intersection distribution:
+  - `S0?! <: S1` if `S0! <: S1`
+
+This is derived from the general rules above as follows:
+  - Given `S0! <: S1`
+  - to show `S0?! <: S1`
+  - that is, to show `(S0 | Null) & Object <: S1`
+  - by transitivity and distributivity it suffices to show that:
+    - `(S0 & Object) | (Null & Object) <: S1`
+  - by left union introduction, it suffices to show that:
+    - `S0 & Object <: S1`, aka `S0! <: S1`
+    - and `Null & Object <: S1`
+  - `S0! <: S1` holds by assumption, so it suffices to show that `Null & Object
+    <: S1`
+  - this holds by the left non-nullable `Null` axiom + transitivity.
+
+
+Right union distribution:
+  - `S0 <: S1!?` if `S0 <: S1?`
+
+This is derived from the general rules above as follows:
+  - Given `S0 <: S1?`
+  - to show `S0 <: S1!?`
+  - that is, to show `S0 <: (S1 & Object) | Null`
+  - by transitivity and distributivity it suffices to show that:
+    - `S0 <: (S1 | Null) & (Object | Null)`
+  - by right intersection introduction, it suffices to show that
+    - `S0 <: (S1 | Null)`
+    - `S0 <: (Object | Null)`
+  - The first holds by assumption
+  - The second holds by the right nullable `Object` axiom
+
+
+The declarative legacy subtyping rules are derived from the possible completions
+of the legacy type to a non-legacy type.
```

Comment:
I would expect that we would have a situation where legacy types should not arise, except as "consistently legacy"; so we wouldn't encounter `List<int>*`, only `List<int*>*`, and we shouldn't encounter `Function(int)*` only `dynamic* Function(int*)*`.

This would presumably also have implications for subtyping: `class IntList extends List<int>` would mean that `IntList*` has `List<int*>*` as a supertype, and similarly for other shapes.